grails-app/controllers/org/pih/warehouse/inventory/StockMovementController.groovy

```diff
@@ -283,6 +286,101 @@ class StockMovementController {
+            return
+        }
+        render(view: "addComment", model: [stockMovement: stockMovement, comment: comment])
+    }
+
+    def deleteComment = {
+        StockMovement stockMovement = stockMovementService.getStockMovement(params['stockMovement.id'])
+        if (!stockMovement) {
+            flash.message = "${g.message(code: 'default.not.found.message', args: [g.message(code: 'stockMovement.label', default: 'StockMovement'), params['stockMovement.id']])}"
+            redirect(action: "list")
+            return
+        }
+        Comment comment = Comment.get(params?.id)
+        if (!comment) {
+            flash.message = "${g.message(code: 'default.not.found.message', args: [g.message(code: 'comment.label', default: 'Comment'), params.id])}"
+            redirect(action: "show", id: stockMovement.id)
+            return
+        }
+
+        if (comment.sender.id != session.user.id) {
+            throw new UnsupportedOperationException("${g.message(code: 'errors.noPermissions.label')}")
+        }
+        stockMovementService.deleteComment(comment, stockMovement.requisition)
+        flash.message = "${g.message(code: 'default.comments.deleted.label')}"
+        redirect(action: "show", id: stockMovement.id)
+    }
+
+    def saveComment = {
+        StockMovement stockMovement = stockMovementService.getStockMovement(params['stockMovement.id'])
+        if (!stockMovement) {
+            flash.message = "${g.message(code: 'default.not.found.message', args: [g.message(code: 'stockMovement.label', default: 'StockMovement'), params['stockMovement.id']])}"
+            redirect(action: "list")
+            return
+        }
+        Comment comment = new Comment(params)
+        stockMovementService.saveComment(comment)
+        stockMovementService.addCommentToRequisition(comment, stockMovement.requisition)
+         if (comment?.hasErrors()) {
+             stockMovement = stockMovementService.getStockMovement(params['stockMovement.id'])
+             render(view: "addComment", model: [stockMovement: stockMovement, comment: comment])
+             return
+         }
+        flash.message = "${g.message(code: 'default.comments.created.label')}"
+        redirect(action: "show", id: stockMovement.id)
+    }
+
+    def updateComment = {
+        StockMovement stockMovement = stockMovementService.getStockMovement(params['stockMovement.id'])
+        if (!stockMovement) {
+            flash.message = "${g.message(code: 'default.not.found.message', args: [g.message(code: 'stockMovement.label', default: 'StockMovement'), params['stockMovement.id']])}"
+            redirect(action: "list")
+            return
+        }
+        Comment comment = Comment.get(params?.id)
+        if (comment.sender.id != session.user.id) {
+            throw new UnsupportedOperationException("${g.message(code: 'errors.noPermissions.label')}")
+        }
+        comment.properties = params
+        stockMovementService.saveComment(comment)
```

Comment:
since comment is a managed entity at this point, it is not necessary to save() it explicitly - we could (like I proposed above), just do:
```groovy
if (comment.validate()) {
  flash.message = "${g.message(code: 'default.comments.updated.label')}"
  redirect(action: "show", id: stockMovement.id)
  return
}
stockMovement = stockMovementService.getStockMovement(params['stockMovement.id'])
render(view: "addComment", model: [stockMovement: stockMovement, comment: comment])
```
but in this place, don't treat is as a "must change request".